### PR TITLE
feat(parser) add flag to log diff on failure

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
Add a `--log-failed-config` flag to write a config diff to disk if applying that config failed.

**Background**
Diagnosing reconciliation loop failures is currently difficult for end users. In the event of a failure, we log admin API responses, but those don't always provide a clear indication of the issue: information about conflicts or invalid configuration isn't always easy to link back to Kubernetes resources, and in some cases, the admin API just encounters a fatal error that provides no useful info at all. Tracebacks or debug messages in Kong logs may provide additional hints, but those aren't necessarily intuitive either.

Logging what the controller _wants_ to apply can simplify that debugging process: if you can see the proposed config, and how it differs from the last successfully-applied configuration, it should be easier to find problem Kubernetes resources related to it or apply it to a test Kong instance that does not have the same Kubernetes resources available--useful since we usually want to avoid shipping and running debug proxy code in actual user environments if possible.

**Implementation considerations**
- Ideally, we'd show a diff per resource to cut down on noise info (we don't care about the config that does work). However, the controller builds config into a single unit before passing it off to the DB-less config endpoint or deck (which can't report back about K8S resources because it has no knowledge of them). Diffing the complete configuration is imperfect, but provides at least some additional information in absence of more granular information about which K8S resources were used to derive generated configuration.
- The complete configuration can contain sensitive information, so this should be gated behind a flag to only dump config when you know you have an issue and want to enable debug dumps. That clashes with the diff approach if you're unsure which piece of K8S config is it fault, so I probably need to have it dump the raw config also.
- Complete config dumps would be noisy in logs, and again, some config is sensitive, so while writing a temporary file that you'll then need to copy out of a running container is a bit cumbersome, it seems better than logging.
- This excludes custom entities, but as those are just a user-provided YAML blob, I don't see much reason to include them--we really only care about stuff sourced from K8S config.
- This borrows deck's JSON diff function with some minor modifications. The controller doesn't touch deck when building a DB-less state file, so we can't reuse that yet--if we're adding that to deck ([as is proposed](https://github.com/Kong/deck/issues/238)), then it'd probably make sense to instead add a report back mechanism to the deck solver and have it provide the diff content.

### CODE REVIEW

**Comment by hbagdi** [comment]
Why can't this be done at `debug` log level instead?
+1 to adding this feature but -1 for adding yet another configuration option.

**Comment by rainest** [comment]
Rework a bit to create a temp dir at start and use that, since the linter doesn't like static paths under /tmp, and that's a realistic possible concern in container-less dev runs.

The rationale for the additional config option is twofold:
- We generate a complete state from all resources and then attempt to apply it, and that attempt may fail. We can't reliably know what part is the problem, since it's either a single blob we pass to `/config` (which may just respond with a "nope! something went wrong!" error, with a corresponding traceback like #884 that doesn't provide a good indication of what failed) or a single blob we pass to deck's solver (which does run individual requests, but doesn't report back details about them, and can also encounter inscrutable Lua tracebacks). The controller will diligently retry on failure, so while we could log the blob on each attempt, it'd be quite verbose and repetitive.
- The state blob can and will contain sensitive info (credentials and such), and we want to limit where it goes. The standard logging channel will often persist elsewhere (scraping container logs and saving them to a central log collector is common practice in many K8S clusters), and we don't want state info to go into it. We can (and should) mitigate that by adding a sanitize mode to `toDeckContent()` (rather, add new deck logic that strips sensitive content and a flag to `toDeckContent()` that enables it), but that still wouldn't make it any less noisy.

Since we may ask users to enable debug logs for problems that don't require full state dumps, I think it's reasonable to add a separate flag for it.

My end goal here is to provide a simpler end-user diagnostic experience while intentionally gating it behind something that you wouldn't normally enable for other reasons. Currently, getting state info out of the controller requires either hooking up a Go debugger (to inspect in-memory state) or a tcpdump sidecar (to inspect state as the controller sends it to the admin API over the network). Those have the same sensitive data concerns alongside usability hurdles, because they're general-purpose tools and require knowledge of how our systems work to extract data from them effectively. "Enable flag, copy contents of temp directory as problem occurs, ship results to us" are hopefully simpler instructions than "eh, just tcpdump it!", and dumping state further gets us something we can easily use to reproduce a problem offline in a test environment (since it's the format that deck and DB-less use).

**Comment by hbagdi** [comment]
To be clear, I'm not against adding this feature itself. This is certainly useful for debugging but I'm not convinced that this needs to be behind a new configuration flag. This should be logged at debug-level. Debug-level can have sensitive info in it and of course, one shouldn't use it in production.

**Comment by rainest** [comment]
In practice, we may need to use this in production sometimes. It's not always possible to replicate an unexplained issue in a test environment. It arguably should be, but the ease of doing so depends on tooling we don't control and is prone to user error. Furthermore, even when replicating production state in a test environment is possible, a perfect replica will still contain production's sensitive values.

Further rework to instead make this part of debug-level logging, but redact certificate keys and credential secrets by default, only showing them if you further pass `--log-sensitive-config`. Still needs testing to validate.

The implementation overwrites the fields in question in state in-place after attempting to apply it. This should be fine as we generate state from scratch on every update anyway, so the mangled data only goes to the diff log.

**Comment by mflendrich** [comment]
On a high level, I have several concerns:
- I'm not a fan of making KIC's logic more complex by adding new execution paths and outputs, which is what this PR does.
- As is, this PR leaks tempdirs at the rate of one per KIC process lifetime. Both in a non-containerized and a containerized (e.g. crash loop) scenario this can lead to uncontrolled storage exhaustion. We have to control the number of files that KIC has potential to create over time.
- The proposed sanitization mechanism is subtractive (and, hence, has potential to miss potentially sensitive entities added in the future, and puts additional burden on developers: requires them to remember that this function exists, and to update it). Anonymization/sanitization is an ungrateful problem (ask anyone who's worked on GDPR compliance :grin:) and it's better to avoid having it at all. From a security perspective, anyone with access to KIC sufficient to manage these settings and read the file, is able to get the sensitive details, so the benefit of implementing sanitization here is unclear to me.

Therefore, the cost-benefit ratio of this change isn't on the undoubtedly positive side for me. I would like to propose a wholly different approach in line with the KISS principle. Not sure how much work it would be - just a thought experiment.

The 30k ft view of how KIC works today is roughly the following (stateless, except for `state.State` changes being incremental) pipeline triggered by the controller loop:
`K8s etcd contents -1-> state.State -2-> KongState -3-> file.Contents -4-> Kong`

The solution would be to refactor appropriate steps (`-1->` through `-4->`), unless done so already, into reusable "filters". We could then implement a one-shot command-line tool (that tool might hypothetically be a `cobra` subcommand of KIC, or a separate binary) that would run a subset of the pipeline just once (as opposed to running in a controller loop). With such a tool, the solution to the subject problem would be:

> oops, KIC is failing for a customer. Ask them to run the one-shot tool, as a k8s pod, executing steps 1, 2, 3 of the pipeline and printing/dumping the result

Having KIC structured like this would additionally help us test part of KIC's functionality in better separation.

**Review by mflendrich — COMMENTED** [review]
On a high level, I have several concerns:
- I'm not a fan of making KIC's logic more complex by adding new execution paths and outputs, which is what this PR does.
- As is, this PR leaks tempdirs at the rate of one per KIC process lifetime. Both in a non-containerized and a containerized (e.g. crash loop) scenario this can lead to uncontrolled storage exhaustion. We have to control the number of files that KIC has potential to create over time.
- The proposed sanitization mechanism is subtractive (and, hence, has potential to miss potentially sensitive entities added in the future, and puts additional burden on developers: requires them to remember that this function exists, and to update it). Anonymization/sanitization is an ungrateful problem (ask anyone who's worked on GDPR compliance :grin:) and it's better to avoid having it at all. From a security perspective, anyone with access to KIC sufficient to manage these settings and read the file, is able to get the sensitive details, so the benefit of implementing sanitization here is unclear to me.

Therefore, the cost-benefit ratio of this change isn't on the undoubtedly positive side for me. I would like to propose a wholly different approach in line with the KISS principle. Not sure how much work it would be - just a thought experiment.

The 30k ft view of how KIC works today is roughly the following (stateless, except for `state.State` changes being incremental) pipeline triggered by the controller loop:
`K8s etcd contents -1-> state.State -2-> KongState -3-> file.Contents -4-> Kong`

The solution would be to refactor appropriate steps (`-1->` through `-4->`), unless done so already, into reusable "filters". We could then implement a one-shot command-line tool (that tool might hypothetically be a `cobra` subcommand of KIC, or a separate binary) that would run a subset of the pipeline just once (as opposed to running in a controller loop). With such a tool, the solution to the subject problem would be:

> oops, KIC is failing for a customer. Ask them to run the one-shot tool, as a k8s pod, executing steps 1, 2, 3 of the pipeline and printing/dumping the result

Having KIC structured like this would additionally help us test part of KIC's functionality in better separation.

**Comment by rainest** [comment]
> I'm not a fan of making KIC's logic more complex by adding new execution paths and outputs, which is what this PR does.

This could be said of any new functionality. This augments `OnUpdate()` and adds a couple simple helper functions to output additional diagnostic information without modifying translation or reconciliation. I think it's reasonable.

> As is, this PR leaks tempdirs at the rate of one per KIC process lifetime.

There shouldn't be more than one process under normal circumstances--`controller.Start()` (which creates the temporary directory) is our container PID 1. If it dies, the container gets recycled.

I'm not really concerned with non-containerized instances since they should only be used for development work.

> so the benefit of implementing sanitization here is unclear to me.
We may ask users to ship these to us for review.

That aside, I do need to rework the redaction method to produce valid configuration: the basic `REDACTED` string will fail validation since it's not unique (credentials and certificates need some random component to not collide) and not a PEM (the certificate endpoint requires an actual certificate with a matching key). Generating better values for those should be pretty simple, however.

> the cost-benefit ratio of this change isn't on the undoubtedly positive side for me.

We don't currently have any good means of diagnosing the cause of a conflict (the single generated Kong state prevents us from logging Kong errors on a per-resource basis) or incorrect translation (the only way to see admin API payloads is to use tcpdump). 

These difficulties are a consequence of our current reconciliation loop, and I agree that we'd be best served by refactoring it. However, we don't have set idea of what that'd look like or when we'd be able to complete it, and with that in mind, I think we need to add something to better diagnose problems in the existing loop. If you want practical examples of where/how this would be useful:

- I was unable to determine the cause (ultimately two consumers with the same username) of a reconciliation failure in a fresh test environment because we don't receive detailed errors from deck (we only receive a generic "config rejected" error) and don't report what we're processing well (when syncing to an empty database, the controller seemingly reports that it's syncing a single, random piece of configuration only, and can even report that it's "syncing" things it will ultimately discard because of a class mismatch). Dumping the target configuration and attempting to apply it using deck indicated the issue instantly, since deck _does_ report individual problem entities when used independently.
- Support and I spent several days working with a user attempting to locate the issue with some configuration. Much of this time was spent working through incorrect hypotheses developed based on inaccurate user reports and piecemeal data (subsets of K8S configuration, PSQL backups, and tcpdumps) that did not allow us to replicate the issue. It'd be much easier to do so if we could say "dump the current DB contents and collect the controller dump", as they're in the same format and can be loaded directly into Kong.
- My attempt to replicate #884 just failed; I don't get the same error and that configuration is loaded as expected. That sort of issue (where we have an unhandled Lua exception) often requires custom Lua code to debug effectively, and it's much easier to develop and  apply that locally than to go back and forth with the user seeing the issue. With a configuration dump, we don't need to worry about determining why an attempt to reproduce K8S configuration isn't causing the same issue.

**Comment by rainest** [comment]
> I do need to rework the redaction method to produce valid configuration: the basic REDACTED string will fail validation since it's not unique (credentials and certificates need some random component to not collide) and not a PEM (the certificate endpoint requires an actual certificate with a matching key)

Having played around with this some, it works, but purely random values always generate a diff for these resources. To avoid that, either:
- Delete the sensitive items altogether. This is fine for credentials (nothing depends on them) but poses a minor challenge for certificates, which the Service `client-cert` annotation can reference (SNIs reference certificates also, but are children of go-kong Certificate objects).
- Deterministically generate the redaction content. For certificates, seed key generation with something (serial?), and for credentials, hash somehow. Either SHA2 with a random salt (chosen at startup) if we don't care about comparisons across instances; bcrypt if we do.

File: internal/ingress/controller/kong.go [hunk 1]
```diff
@@ -72,10 +93,24 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+	var target []byte
+	if n.cfg.LogSensitiveConfig {
+		target, _ = json.Marshal(targetContent)
+	} else {
+		state.Sanitize()
+		sanitizedContent := n.toDeckContent(ctx, state)
+		target, _ = json.Marshal(sanitizedContent)
+	}
 	if err != nil {
+		if n.cfg.LogLevel == "debug" {
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:105]: I believe (in disagreement with @hbagdi's https://github.com/Kong/kubernetes-ingress-controller/pull/991#issuecomment-749663141) that this functionality (of adding a new KIC output) should not be correlated with logging in any way because:
- [12factor](https://12factor.net/logs) says the following about logging:
    > A twelve-factor app never concerns itself with routing or storage of its output stream. It should not attempt to write to or manage logfiles. Instead, each running process writes its event stream, unbuffered, to `stdout`.
    
    Any attempt at being concerned with where and how the logs are stored is a violation of 12factor.
- It's an anti-pattern for containers to log to the local filesystem
- All features (other than being likely useful with a detailed logging level) of this functionality are orthogonal to the log stream produced by the application.
- While logging is a standard, non-unique operational concern (like metrics, GC, etc.) providing a dump from the business logic is not. Therefore I'm tempted to consider dumping the diff an auxiliary feature of the application.

Therefore, I propose the following flag set to control this feature:
```go
flag.String("failure-dump-dir", "", "If set to a directory, the ingress controller will, in case of failed application, save the generated deck configuration and the diff to the specified directory.")
flag.Bool("failure-dump-sanitized", true, "If true, the ingress controller will do the best effort to redact sensitive contents from the files saved to --failure-dump-dir.")
```

This solution brings the following benefits:
- the user specifying a destination dir enables the feature (no need to hardcode the temp dir)
- it's impossible to accidentally spam the local FS by changing log level

File: internal/ingress/controller/kong.go [hunk 2]
```diff
@@ -72,10 +93,24 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+			diff, _ := getDiff(target, n.lastConfig)
+			_ = ioutil.WriteFile(n.tmpDir+"/target.json", target, 0600)
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:107]: Use `filepath.Join` to join path segments.

File: internal/ingress/controller/parser/kongstate/consumer.go [hunk 3]
```diff
@@ -24,6 +24,25 @@ type Consumer struct {
...
+		auth.Secret = &redacted
+	}
+	for _, auth := range c.BasicAuths {
+		auth.Password = &redacted
+	}
+	for _, auth := range c.Oauth2Creds {
+		auth.ClientSecret = &redacted
+	}
+}
+
```
Comment (mflendrich) [inline internal/ingress/controller/parser/kongstate/consumer.go:45]: Two independent thoughts here.

1. #822 made the `parser.Build()` function a pure function (in a sense that it takes an argument, `store.Storer`, and produces a value, `kongstate.KongState` without any side effects other than logging). I believe there are [strong benefits ](http://www.javapractices.com/topic/TopicAction.do?Id=29)(which I'm happy to discuss further in disconnect from this PR, if need be) of making the logic in this part of the codebase functional. I consider the work done in #822 (making `parser.Build` a pure function, and turning all functions that mutate an object into member functions of that object) a step in that direction. Writing new functions is IMO a good opportunity to advance this effort - by writing not mutator functions, but pure functions that return an object if changed.
2. In line with my other comment: this implementation of redaction is subtractive (takes an entire object and wipes away some closed list of fields) which is an anti-pattern for redaction, because addition of a new field and forgetting to update the list is guaranteed to result in a leak.

Therefore, PTAL at #1021 which is an implementation of `kongstate.SanitizedCopy()` which is additive (only copies fields which are known to be safe) and refactors a further part of the codebase in line with the considerations above.

File: internal/ingress/controller/kong.go [hunk 4]
```diff
@@ -72,10 +93,24 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+	if n.cfg.LogSensitiveConfig {
+		target, _ = json.Marshal(targetContent)
+	} else {
+		state.Sanitize()
+		sanitizedContent := n.toDeckContent(ctx, state)
+		target, _ = json.Marshal(sanitizedContent)
+	}
 	if err != nil {
+		if n.cfg.LogLevel == "debug" {
+			diff, _ := getDiff(target, n.lastConfig)
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:106]: FWIW, [existing CLI tools can diff two JSONs](https://stackoverflow.com/questions/31930041/using-jq-or-alternative-command-line-tools-to-compare-json-files). Adding extra diffing functionality here may end up being superfluous, unless CLI diffing is a significant drag.

I understand that the UX with a diff produced directly by KIC is more streamlined, but I don't see how it would justify the extra complexity added to the codebase.

**Review by mflendrich — CHANGES_REQUESTED** [review]
As agreed in a video call (@rainest @mflendrich @shaneutt):
- the long-term solution to this problem should be a standalone tool that performs one pass of KIC and produces intermediate results
- as a stop-gap (because the long-term solution is more costly) we'll implement config dumping as a side effect of the reconciliation loop (which this PR does).

PTAL at the attached comments (written with a motivation to minimize the added tech debt associated with the stop-gap)

File: internal/ingress/controller/parser/kongstate/consumer.go (hunk 3 above)
Comment (rainest) [reply internal/ingress/controller/parser/kongstate/consumer.go]: What was the purpose of the accompanying `SetCredential()` refactor? I think that may no longer be necessary with the updated strategy in c9ab6de, which drops sensitive items entirely and removes references to them, since it no longer modifies credentials, and it looks like the goal there was to make that modification a function of the credentials rather than the consumers. Assuming so, I think 546a81c to build a copy from chosen fields without that refactor is fine.

I'd originally wanted to produce a redacted version that had credentials and certificates to clearly indicate when the controller wanted to modify them while not disclosing their actual values. Generating appropriate fake values that still exhibited changes when appropriate proved more trouble than it was worth, so I've opted to just drop them entirely.

File: internal/ingress/controller/kong.go (hunk 1 above)
Comment (rainest) [reply internal/ingress/controller/kong.go]: I don't particularly care what we call it. da2c974 makes it a string flag that lets you select whether it includes sensitive data.

Do you indeed think writing to the filesystem is indeed likely to cause issues in practice? Again, this can't create more than one set of dump files because it's tied to the container master process, and if that dies, the container filesystem dies with it. I don't see how choosing a location makes any difference unless we expect user to mount a dedicated volume for it. What we're writing here doesn't warrant that overhead.

A possibly better solution would be to provide a download endpoint on [the status server](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.1.1/cli/ingress-controller/main.go#L647-L714), but we don't have any way to expose data from the sync loop to it, and I don't think this warrants building one.

Dumping to the filesystem isn't perfect, but it's simple and should be sufficient for this.

Comment (mflendrich) [reply internal/ingress/controller/kong.go]: > I don't particularly care what we call it

My point is about separating these two concerns:
- logging (which is well defined in cloud-native best practices, and the subject feature does IMO clearly not belong)
- writing intermediate results to the FS

Naming is merely a consequence of being careful about separation of concerns (vide: "ubiquituous language" in DDD).

Why I think it's important to maintain these concerns as separate:
- "Logging", like I said before, is well defined for cloud-native applications. Specifically, raising the logging level is widely expected to increase the volume of stdout, not to produce extra files. Sure, we can nurture the "tribal knowledge" that "KIC is special in that setting the log level to debug has this and that side effect", but such things add up over time as technical debt.
- Doing otherwise is at odds with several points of SOLID.
- Implementing this feature separately from logging limits points of integration between this feature and the rest of the system to the `KongState` and `file.Content` objects. Keeping the size of the integration minimal is highly desirable. Doing otherwise couples this feature with logging, which increases the size of the integration.

> Do you indeed think writing to the filesystem is indeed likely to cause issues in practice?

If we make writing to the filesystem a side effect of increasing the logging level, I think it will cause the following issues:
- violate the cloud-native contract of what logging is (vide the 12factor link above)
- introduce a non-0 probability of data leak and full disk issues
- make logging and the control loop harder to maintain

> Again, this can't create more than one set of dump files because it's tied to the container master process, and if that dies, the container filesystem dies with it.

This is not incorrect, but it depends on two further unrelated facts:
- that the Dockerfile puts KIC as PID 1 in the container,
- that Kubernetes container lifecycle deletes and recreates a container on restart (unlike docker)

Not depending on this behavior removes a whole class of problems.

> I don't see how choosing a location makes any difference unless we expect user to mount a dedicated volume for it

Making environment-specific parameters external is another principle of [12factor](https://12factor.net/config). Rather than speculating whether the user will use this in a volume or not, making this external effectively removes the question from our list of worries altogether.

> A possibly better solution would be to provide a download endpoint on the status server

I think that writing to the FS is just fine - it's simple and solves the problem. What makes a difference, though, is _how_ we implement that (from a software architecture standpoint). Which is my key point in this review.

To summarize all the points above, my distilled requests are:
- Treat generated config dumping as functionality that's minimally related to other areas of concern in KIC (such as logging).
- If there's no good reason to do otherwise, follow general SW design guidance such as SOLID, 12factor.
- Minimize interactions between different areas of concern (some examples given above)

I believe that the proposal from https://github.com/Kong/kubernetes-ingress-controller/pull/991#discussion_r564687599 is an easy way to achieve that.

File: internal/ingress/controller/parser/kongstate/consumer.go (hunk 3 above)
Comment (mflendrich) [reply internal/ingress/controller/parser/kongstate/consumer.go]: > What was the purpose of the accompanying `SetCredential()` refactor?

Unrelated to the gist of `SanitizedCopy()`. The history behind the `SetCredential()` refactor is the following:
Implementing `SanitizedCopy()` required `KeyAuth`, `HMACAuth`, et al. to implement `SanitizedCopy()`. This was not possible without these types being defined in KIC's codebase. Defining these types created an (unrelated) opportunity to shift the business logic of individual auth types from a "god function" (which `SetCredential` was before the refactor) into specialized "classes".

In short: there's no reason for the `SetCredential()` refactor to be a part of #1021, other than this refactor being a drive-by architecture fix that was easy to carry out given this change. I think that the `SetCredential()` refactor makes sense regardless of the course we take with `SanitizedCopy()`. If you wish, I'm happy to extract the `SetCredential` refactor to a PR unrelated to #1021.

> it looks like the goal there was to make that modification a function of the credentials rather than the consumers

There were two goals:
1. replace mutating an existing `KongState` or any of its subentities with creating a fresh copy (to achieve object immutability)
2. replace "erasing fields known to be secret" (that is, taking KongState as a whole and wiping off cherry-picked fields) with "include only fields known to be non-secret". Slightly different logic, yet better behaved in failure cases.

546a81c does not fulfill the goal number 2.

To summarize: I'd like use to come up with a solution that achieves (1) and (2). #1021 appears to me as such. You seem to be raising some further problems with that solution which are not clear to me.

File: cli/ingress-controller/flags.go [hunk 5]
```diff
@@ -311,6 +319,9 @@ func parseFlags() (cliConfig, error) {
 	config.LogLevel = viper.GetString("log-level")
 	config.LogFormat = viper.GetString("log-format")
 
+	// Diagnostics
+	config.DumpConfig = viper.GetString("dump-config")
```
Comment (mflendrich) [inline cli/ingress-controller/flags.go:323]: `cfg.DumpConfig` _behaves like_ an enumeration type, and I think it'd make sense to implement it as such:

```go
type ConfigDumpMode int

const (
	ConfigDumpModeOff ConfigDumpMode = iota
	ConfigDumpModeEnabled ConfigDumpMode = iota
	ConfigDumpModeSensitive ConfigDumpMode = iota
)
```

with a function like `func parseConfigDumpMode(in string) (ConfigDumpMode, error)` being run here, to parse the `string` flag into a `DumpConfigMode`, and failing (exiting) if the value doesn't parse.

File: internal/ingress/controller/controller.go [hunk 6]
```diff
@@ -231,12 +235,21 @@ type KongController struct {
...
 // Stop.
 func (n *KongController) Start() {
 	n.Logger.Debugf("startin up controller")
+	if len(n.cfg.DumpConfig) > 0 {
+		var err error
+		n.tmpDir, err = ioutil.TempDir("", "controller")
+		if err != nil {
+			panic(fmt.Errorf("failed to create a temporary working directory: %v", err))
+		}
+	}
```
Comment (mflendrich) [inline internal/ingress/controller/controller.go:252]: There are many similar initializations in the body of `main()`. I recommend moving this block there.

File: internal/ingress/controller/controller.go [hunk 7]
```diff
@@ -231,12 +235,21 @@ type KongController struct {
...
 
 // Start starts a new master process running in foreground, blocking until the next call to
 // Stop.
 func (n *KongController) Start() {
 	n.Logger.Debugf("startin up controller")
+	if len(n.cfg.DumpConfig) > 0 {
+		var err error
+		n.tmpDir, err = ioutil.TempDir("", "controller")
+		if err != nil {
+			panic(fmt.Errorf("failed to create a temporary working directory: %v", err))
```
Comment (mflendrich) [inline internal/ingress/controller/controller.go:250]: the codebase uses logger's `Fatalf` for this purpose. I recommend changing to that (instead of panicking) for consistency.

Also, `Start()` doesn't seem designed for failures (and it's not easy to cleanly return an error from here) - which reinforces my point about moving this to `main()`.

File: internal/ingress/controller/kong.go [hunk 8]
```diff
@@ -72,10 +74,22 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+	if n.cfg.DumpConfig == "enabled" {
+		target, _ = json.Marshal(targetContent)
+	} else if n.cfg.DumpConfig == "sensitive" {
+		sanitizedState := state.SanitizedCopy()
+		target, _ = json.Marshal(n.toDeckContent(ctx, sanitizedState))
+	}
 	if err != nil {
+		if len(target) > 0 {
+			_ = ioutil.WriteFile(filepath.Join(n.tmpDir, "target.json"), target, 0600)
+			_ = ioutil.WriteFile(filepath.Join(n.tmpDir, "last_good.json"), n.lastConfig, 0600)
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:87]: I think that these errors would be worth logging on the `Warn` level.

File: internal/ingress/controller/kong.go [hunk 9]
```diff
@@ -72,10 +74,22 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+		}
 		return err
 	}
 	n.runningConfigHash = shaSum
+	n.lastConfig = target
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:92]: Would this way to write "last good config" and "last bad config" work as well:

1. try to apply config
2. if application was successful, write to "last_good.json"
3. otherwise, write to "last_failed.json"?

This would free us of the need to remember `lastConfig`.

File: internal/ingress/controller/controller.go [hunk 10]
```diff
@@ -231,12 +235,21 @@ type KongController struct {
 	PluginSchemaStore PluginSchemaStore
 
 	Logger logrus.FieldLogger
+
+	tmpDir string
```
Comment (mflendrich) [inline internal/ingress/controller/controller.go:239]: This is invalid if `DumpConfig` is not enabled. Therefore, the name of this field should be closely tied to that functionality, and there should be a comment saying in which circumstances this field is usable.

Something along the lines of

```go
// dumpConfigDir is the target directory for config dumps, if dumping is enabled.
dumpConfigDir string
```

File: cli/ingress-controller/main.go [hunk 11]
```diff
@@ -453,8 +455,17 @@ func main() {
 
 	store := store.New(cacheStores, cliConfig.IngressClass, cliConfig.ProcessClasslessIngressV1Beta1,
 		cliConfig.ProcessClasslessIngressV1, cliConfig.ProcessClasslessKongConsumer, log.WithField("component", "store"))
+
+	var dumpDir string
+	if cliConfig.DumpConfig > 0 {
```
Comment (mflendrich) [inline cli/ingress-controller/main.go:460]:
```suggestion
	if cliConfig.DumpConfig != util.DumpConfigOff {
```

File: cli/ingress-controller/flags.go [hunk 12]
```diff
@@ -311,6 +320,9 @@ func parseFlags() (cliConfig, error) {
 	config.LogLevel = viper.GetString("log-level")
 	config.LogFormat = viper.GetString("log-format")
 
+	// Diagnostics
+	config.DumpConfig = util.ParseConfigDumpMode(viper.GetString("dump-config"))
```
Comment (mflendrich) [inline cli/ingress-controller/flags.go:324]: Neither this flag, nor `util.ParseDumpConfig`, is tested now. I think this deserves a test in `flag_test.go`, involving all possible values of the `dump-config` flag (`"sensitive"`, `"enabled"`, `""`, `"sdifohjwoeui"`)

File: cli/ingress-controller/main.go [hunk 13]
```diff
@@ -453,8 +455,17 @@ func main() {
...
+		var err error
+		dumpDir, err = ioutil.TempDir("", "controller")
+		if err != nil {
+			log.Fatalf("failed to create a dump directory: %v", err)
+		}
+	}
 	kong, err := controller.NewKongController(ctx, &controllerConfig, updateChannel,
-		store)
+		store, dumpDir)
```
Comment (mflendrich) [inline cli/ingress-controller/main.go:468]: `NewKongController` takes a lot of arguments, which are grouped in `controllerConfig`. Why add a new positional argument instead of extending `controllerConfig`?

File: internal/ingress/controller/kong.go [hunk 14]
```diff
@@ -72,8 +74,28 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
...
+				n.Logger.Warnf("failed to dump configuration: %s", dumpErr)
+			}
+		}
 		return err
+	} else {
+		if len(target) > 0 {
+			dumpErr := ioutil.WriteFile(filepath.Join(n.dumpDir, "last_good.json"), target, 0600)
+			if dumpErr != nil {
+				n.Logger.Warnf("failed to dump configuration: %s", dumpErr)
+			}
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:97]: These two branches (`err != nil` and `else`) differ only by a string literal. There's no need to duplicate the logic because of that. How about creating a helper function:
```go
func buildDumpPath(runFailed bool, dumpDir string) {
	if runFailed {
		return filepath.Join(dumpDir, "target.json") // I don't understand the thought behind the name "target.json" but let's not bikeshed
	}
	return filepath.Join(dumpDir, "last_good.json")
}
```

and replacing both branches with this `if` with something streamlined, for example:
```go
dumpPath := buildDumpPath((err != nil), n.dumpDir)
if dumpErr := ioutil.WriteFile(dumpPath, target, 0600); dumpErr != nil {
	n.Logger.WithField("dump_path", dumpPath).WithError(dumpErr).Warnf("failed to dump configuration)
}
```
?

File: internal/ingress/controller/kong.go [hunk 15]
```diff
@@ -72,8 +74,28 @@ func (n *KongController) OnUpdate(ctx context.Context, state *kongstate.KongStat
 	} else {
 		err = n.onUpdateDBMode(targetContent)
 	}
+	var target []byte
+	if n.cfg.DumpConfig == util.ConfigDumpModeSensitive {
+		target, _ = json.Marshal(targetContent)
+	} else if n.cfg.DumpConfig == util.ConfigDumpModeEnabled {
+		sanitizedState := state.SanitizedCopy()
+		target, _ = json.Marshal(n.toDeckContent(ctx, sanitizedState))
```
Comment (mflendrich) [inline internal/ingress/controller/kong.go:82]: These `errors`s should not be ignored because `target` is invalid if `err != nil`.

Instead of ignoring error (and `if len(target) > 0` below), let's have `err != nil` here suppress writing.

I think that this can be cleanly achieved by wrapping the whole new dumping logic being introduced here in a function like `func dumpConfig(configDumpMode, targetContent, dumpDir) error`

**Comment by rainest** [comment]
Variant that incorporates #1029 in https://github.com/Kong/kubernetes-ingress-controller/tree/feat/dump-config pending that getting merged to next.

**Review by mflendrich — APPROVED** [review]
